src/pattern_matching/myers.rs

```diff
@@ -156,29 +446,374 @@ impl<'a, I: Iterator<Item = &'a u8>> Iterator for Matches<'a, I> {
+
+    /// Updates the given `Alignment` with its position
+    /// and alignment path if found. The distance is stored in `Alignment::score`.
+    /// See `start()` for a description of corner cases.
+    #[inline]
+    pub fn alignment(&mut self, aln: &mut Alignment) {
+        let (len, _) = self.myers.tb.traceback(Some(&mut aln.operations));
+        update_aln_positions(self.pos, len, self.state.dist, self.m, aln);
+    }
+}
+
+
+impl<'a, T, I, S> Iterator for FullMatches<'a, T, I, S>
+where T: 'a + Num,
+      I: TextIterator<'a>
+{
+    type Item = (usize, usize, usize);
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next_end()
+            .map(|(end, dist)| (self.start(), end + 1, dist))
+    }
+}
+
+
+impl<'a, T, I> FullMatches<'a, T, I, Remember>
+where T: 'a + Num,
+      I: TextIterator<'a>,
+{
+    /// Takes the end position of a hit and returns a tuple of the corresponding starting position
+    /// and the hit distance. If the end position is greater than the end position of the previously
+    /// returned hit, `None` is returned.
+    /// *Note:* A 0-based end position is expected (as returned by `next_end`).
+    pub fn hit_at(&self, end_pos: usize) -> Option<(usize, usize)> {
+        self.myers.tb.traceback_at(end_pos, None)
+            .map(|(len, dist)| (end_pos + 1 - len, dist))
+    }
+
+    /// Takes the end position of a hit and returns a tuple of the corresponding starting position
+    /// and the hit distance. The alignment path is added to `ops`.
+    /// As in `hit_at`, the end position has to be searched already, otherwise `None` is returned.
+    pub fn path_at(&self, end_pos: usize, ops: &mut Vec<AlignmentOperation>) -> Option<(usize, usize)> {
+        self.myers.tb.traceback_at(end_pos, Some(ops))
+            .map(|(len, dist)| (end_pos + 1 - len, dist))
+    }
+
+    /// Takes the end position of a hit and returns a tuple of the corresponding starting position
+    /// and the hit distance. The alignment `aln` is updated with the position, alignment path
+    /// and distance (stored in `Alignment::score`).
+    /// If the end position has not yet been searched, nothing is done and `false` is returned.
+    pub fn alignment_at(&self, end_pos: usize, aln: &mut Alignment) -> bool {
+        if let Some((aln_len, dist)) = self.myers.tb.traceback_at(end_pos, Some(&mut aln.operations)) {
+            update_aln_positions(end_pos, aln_len, dist, self.m, aln);
+            return true;
+        }
+        false
+    }
+}
+
+// Assumes *0-based* end positions, the coordinates will be converted to 1-based
+#[inline(always)]
+fn update_aln_positions(end_pos: usize, aln_len: usize, dist: usize, m: usize, aln: &mut Alignment) {
+    aln.xstart = 0;
+    aln.xend = m;
+    aln.xlen = m;
+    aln.ylen = aln_len;
+    aln.yend = end_pos + 1;
+    aln.ystart = aln.yend - aln.ylen;
+    aln.mode = AlignmentMode::Semiglobal;
+    aln.score = dist as i32;
+}
+
+
+struct Traceback<T: Num> {
```

Comment:
The traceback for Myers is very interesting. Have you invented that yourself or is it from a textbook? Could you add details about the complexity implications? How is it configured?

Reply:
I experimented with different implementations, but ended up doing it [similarly as Edlib does](https://github.com/Martinsos/edlib/blob/6967b806275cd4766ed6c58342a0dff8509e70ad/edlib/src/edlib.cpp#L924). Of course, the implementation was much simpler because I didn't have to care about blocks, and only semiglobal alignment possible, while Edlib also does global and local alignments. The author states that his implementation is inspired by [SSWL](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0082138), but I haven't looked into this. I was searching for other implementations, also in SeqAn, but didn't find any, though it's possible that there is one.

There is some description in [this master thesis; chapter 4.2.2 (pg. 32)](https://bib.irb.hr/datoteka/758607.diplomski_Martin_Sosic.pdf). There is also a [paper about Edlib](https://academic.oup.com/bioinformatics/article/33/9/1394/2964763), but there I couldn't find any details about the implementation.

Finding the score in the i-th cell always requires checking all bits starting from the last cell down to the i-th cell. The simples approach is to use a loop, but I found that using popcnt (`count_ones()`) was faster, so that's my only contribution :).

`_traceback_at()` starts at the given end position (which should have been searched, otherwise the result will be nonsense) and iterates backwards, finding the scores of individual cells and the best alignment path.

Due to a lack of theoretical knowledge, I'm not completely sure about the complexity implications. There is one loop going from zero to `m` (pattern length), so I guess time complexity could be O(m). Space complexity depends on whether all states are stored ('remembered') or not. In the first case, I guess it's O(n + k), in the second case O(m + k).

I hope this answers your questions, otherwise I'm happy to give more details.

Reply:
Yes, thanks! Please add the used literature to the doc comment.